### PR TITLE
Feature user boards

### PR DESCRIPTION
**Pull request by melonmanchan**
Mega big commit!

### CODE REVIEW

**Comment by n4sjamk-docker** [comment]
Can one of the admins verify this patch?

File: config/index.js [hunk 1]
```diff
@@ -5,8 +5,13 @@ module.exports = {
         opts: {
             safe: true
         },
-        url:     process.env.MONGODB_URL     || 'mongodb://localhost/test',
-        timeout: process.env.MONGODB_TIMEOUT || 5000
+        url:             process.env.MONGODB_URL      || 'mongodb://localhost/test',
```
Comment (Temetz) [inline config/index.js]: Check this mongo stuff.

File: config/schemas/session.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    ret.id = doc.id;
+
+    delete ret._id;
+    delete ret.__v;
+}
+
+/**
+ * BUG See 'config/schemas/board.js' for details.
+ */
+SessionSchema.options.toObject.transform = SessionSchema.options.toJSON.transform;
```
Comment (Temetz) [inline config/schemas/session.js]: missing newline

File: config/strategies/basic.js [hunk 3]
```diff
@@ -3,19 +3,20 @@
 var utils = require('../../utils');
 
 var User          = require('mongoose').model('user');
-var LocalStrategy = require('passport-local').Strategy;
+var BasicStrategy =	require('passport-http').BasicStrategy;
```
Comment (Temetz) [inline config/strategies/basic.js]: Dem tabs break stuff

File: config/strategies/basic.js [hunk 4]
```diff
@@ -3,19 +3,20 @@
...
 
 // Defines the 'req.body' fields used to look for credentials.
 var opts = {
 	usernameField: 'email',
-	passwordField: 'password'
+	passwordField: 'password',
+
```
Comment (Temetz) [inline config/strategies/basic.js]: Extra line

File: config/strategies/basic.js [hunk 5]
```diff
@@ -32,12 +33,13 @@ module.exports = new LocalStrategy(opts, function(email, password, done) {
...
-			return done(null, {
+			return done(null, { 
 				id:       user.id,
 				type:     'user',
-				username: user.email
+				username: user.name,
+				boards:   user.boards
 			});
 		});
+		//return res.redirect('/auth/:provider/callback');
```
Comment (Temetz) [inline config/strategies/basic.js]: Why here is code commented?

File: config/strategies/github.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+module.exports = new GitHub.Strategy(options, function(at, rt, profile, done) {
+        // Strategies for different providers should always return a set format of
+        // the user's profile. This profile will then be added...
+        var user = {
+                name:   profile.username,
+                avatar: profile._json.avatar_url
+        }
+        return done(null, user);
+}));
+
```
Comment (Temetz) [inline config/strategies/github.js]: too many newline

File: config/strategies/google.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+var User           =    require('mongoose').model('user');
+var GoogleStrategy =	require('passport-google-oauth2').Strategy;
+
+// Defines the 'req.body' fields used to look for credentials.
+var opts = {
+	clientID: '161571982407-o698t9ofu4nl56efcu3dkl2f2nftb5du.apps.googleusercontent.com',
+	clientSecret: 'Dd0Me0lL3HT4k8vCdMfvBXBa',
+    callbackURL: 'http://localhost:9002/api/auth/login/callback',
+    realm: 'http://localhost:9002/',
+    token: 'aasi',
```
Comment (Temetz) [inline config/strategies/google.js]: Itse olet

File: config/strategies/google.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+var utils = require('../../utils');
+
+var User           =    require('mongoose').model('user');
+var GoogleStrategy =	require('passport-google-oauth2').Strategy;
+
+// Defines the 'req.body' fields used to look for credentials.
+var opts = {
+	clientID: '161571982407-o698t9ofu4nl56efcu3dkl2f2nftb5du.apps.googleusercontent.com',
+	clientSecret: 'Dd0Me0lL3HT4k8vCdMfvBXBa',
```
Comment (Temetz) [inline config/strategies/google.js]: Its secret you know?

File: middleware/relation.js [hunk 9]
```diff
@@ -4,9 +4,31 @@ var _     = require('lodash');
...
+
+		if (req.resolved.board.members) {
+			req.resolved.board.members.map(function (item) {
+				if (item.user == req.user.id && item.role == 'admin') {
+					hasAccess = true;
+					return isUser && hasAccess;
+				}
+			})
+		}
+
```
Comment (Temetz) [inline middleware/relation.js]: extra newline

File: routes/auth.js [hunk 10]
```diff
@@ -51,37 +60,146 @@ Router.route('/auth/login')
...
+	if(req.query.state) {
+		// passing in 'state' here, from the OAuth request as seen above can be
+		// seen as attempting to link an account to an existing one, so we do
+		// just that...
+
+			User.findOne({ _email: req.account.email }, function(err, user) { 
+				if(err) {
+				return next(utils.error(500, err));
+				}
+
```
Comment (Temetz) [inline routes/auth.js]: \n

File: routes/auth.js [hunk 11]
```diff
@@ -51,37 +60,146 @@ Router.route('/auth/login')
...
+            });
+        } else {
+			// we find the user based on the account somehow, probably 'email' or
+			// something similar is used to make sure we get the right guy
+			User.findOne({ _email: req.account.email }, function(err, user) {
+				if(err) {
+				return next(utils.error(500, err));
+			}
+
```
Comment (Temetz) [inline routes/auth.js]: \n

File: routes/auth.js [hunk 12]
```diff
@@ -89,34 +207,46 @@ Router.route('/auth/logout')
...
+		req.body.username ? username = req.body.username : username = req.body.email;
+
+		new User({ name:      username,
+			       account_type: 'standard',
+			       providers: {
+						basic: {
+								email:   req.body.email,
+								password:req.body.password
+						       }
+				   },
```
Comment (Temetz) [inline routes/auth.js]: tab tab

File: routes/board.js [hunk 13]
```diff
@@ -73,8 +75,9 @@ Router.route('/boards')
 	 */
 	.post(middleware.authenticate('user'))
 	.post(function(req, res, next) {
-		var payload           = req.body;
-		    payload.createdBy = req.user.id;
+		var payload           			 = req.body;
+		payload.members                  = [];
+		    //payload.members[req.user.id] = 'admin';
```
Comment (Temetz) [inline routes/board.js]: this comment could be removed.

File: routes/board.js [hunk 14]
```diff
@@ -38,13 +39,14 @@ Router.route('/boards')
...
 			// Guests can only see the board they have access to...
-			boardQuery = Board.find({ _id: req.user.access });
+			boardQuery = Board.find({ _id: req.user.access }).populate('members.user');
 		}
 		else {
-			// Normal users see the boards they have created.
-			boardQuery = Board.find({ createdBy: req.user.id });
+			// Normal users see the boards they are member or admin to.
+			boardQuery = Board.find({'members.user': req.user.id }).populate('members.user');
+
```
Comment (Temetz) [inline routes/board.js]: \n

File: routes/board.js [hunk 15]
```diff
@@ -73,8 +75,9 @@ Router.route('/boards')
 	 */
 	.post(middleware.authenticate('user'))
 	.post(function(req, res, next) {
-		var payload           = req.body;
-		    payload.createdBy = req.user.id;
+		var payload           			 = req.body;
```
Comment (Temetz) [inline routes/board.js]: tab tab tab

File: routes/board.js [hunk 16]
```diff
@@ -85,30 +88,27 @@ Router.route('/boards')
...
+
 		new Board(payload).save(function(err, board) {
 			if(err) {
 				return next(utils.error(400, err));
 			}
 
-			Board.populate(board, 'createdBy', function(err, board) {
-				if(err) {
-					return next(utils.error(500, err));
-				}
```
Comment (Temetz) [inline routes/board.js]: This is extra

File: routes/board.js [hunk 17]
```diff
@@ -85,30 +88,27 @@ Router.route('/boards')
...
+							'type':     req.user.type,
+							'username': req.user.username,
+						}
+					}).save(function(err) {
+						if(err) return console.error(err);
+					});
 
-				return res.json(201, board);
-			});
+					return res.json(201, board);
```
Comment (Temetz) [inline routes/board.js]: so many tab

File: routes/board.js [hunk 18]
```diff
@@ -721,44 +758,221 @@ Router.route('/boards/:board_id/access/:code')
...
+									}
+									if(err.name == 'MongoError' && err.code == 11000) {
+										return next(utils.error(409, 'Creating new session failed'));
+									}
+									return next(utils.error(500, err));
+								}
+								return res.set('x-access-token', guestToken).json(200, payload);
+							});
+
+
```
Comment (Temetz) [inline routes/board.js]: linebreaks

File: server.js [hunk 19]
```diff
@@ -39,7 +40,9 @@ app.use(passport.initialize());
 
 // Attach 'CORS' middleware to every route.
 app.all('*', require('cors')({
-	exposedHeaders: ['x-access-token']
+	exposedHeaders: ['x-access-token', 'X-Requested-With'],
+	headers: ['Content-Type', 'Authorization', 'X-Requested-With'],
+	preflightContinue: 'localhost:8000/login/callback'
```
Comment (Temetz) [inline server.js]: this localhost thing most likely doesn't work like intended

File: utils/boardedit.js [hunk 20]
```diff
@@ -42,4 +45,10 @@ module.exports = function(req, board, old) {
             emitter.to(board.id).emit('board:event', ev.toObject());
         });
 
+}
+
+function resolveBoardMembers(board) {
+    Board.findById(board.id, function(err, board) {
+        return board.members;
+    })
 }
```
Comment (Temetz) [inline utils/boardedit.js]: Missing newline

File: utils/boardmembers.js [hunk 21]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Created by matti on 26.6.2015.
```
Comment (Temetz) [inline utils/boardmembers.js]: And checked by Teemu... Remove?

File: utils/boardmembers.js [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+    var boardQuery = Board.findById(boardId).populate('members.user');
+
+    boardQuery.exec(function(err, board) {
+        if(err) {
+            return null;
+        }
+        return board.members
+    });
+
+}
```
Comment (Temetz) [inline utils/boardmembers.js]: You also are missing a newline

File: utils/index.js [hunk 23]
```diff
@@ -6,3 +6,5 @@ module.exports.emitter              = require('./emitter');
 module.exports.ticketClamper        = require('./clamper');
 module.exports.sanitize             = require('./sanitize');
 module.exports.export               = require('./export');
+module.exports.resolveBoardMembers  = require('./boardmembers');
+module.exports.pollBoardActivity    = require('./pollboardactivity');
```
Comment (Temetz) [inline utils/index.js]: Newline also

File: utils/pollboardactivity.js [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+
+					emitter.to(savedboard.id).emit('board:event', ev.toObject());
+						});
+					})
+				}
+			})
+
+			}
+	})
+}
```
Comment (Temetz) [inline utils/pollboardactivity.js]: Newline

**Comment by Temetz** [comment]
Looks ok now.

File: routes/auth.js [hunk 25]
```diff
@@ -51,37 +60,146 @@ Router.route('/auth/login')
...
+        // note that the token is also created inside the 'createSession' method
+        // in this pseudo implementation
+        var session = user.sessions.push( createSession(user).id );
+ 
+        // finally we redirect the user back to the client
+      	return res.redirect(RedirectURL + '?access_token=' + newtoken);
+	}
+});
+
+
```
Comment (Temetz) [inline routes/auth.js]: extra line here

**Comment by melonmanchan** [comment]
Great job matti!!!! :full_moon_with_face:

**Pull request merged**